### PR TITLE
Intercepting react api

### PR DESCRIPTION
**Pull request by reshadi**
This is generic code to enable intercepting react api.
(Most of the code if migrated from Flowjs internal repo to here)

### CODE REVIEW

File: packages/hyperion-autologging/src/Surface.ts [hunk 1]
```diff
@@ -0,0 +1,341 @@
...
+   * sure that is not added to the DOM tree.
+   * The following code blocks the effect of `setAttribute` by returning true
+   */
+  IElement.setAttribute.onArgsObserverAdd(function (
+    this: Element,
+    attrName,
+    attrValue,
+  ) {
+    if (
+      attrName === domSurfaceAttributeName &&
```
Comment (schwartzmx) [inline packages/hyperion-autologging/src/Surface.ts:153]: should we consider the `SurfaceDOMString.toString() !== ''`?

File: packages/hyperion-autologging/src/SurfaceConsts.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+/**
+ * Copyright (c) Meta Platforms, Inc. and affiliates. All Rights Reserved.
+ */
+
+'use strict';
+
+export const AUTO_LOGGING_SURFACE = "data-surface";
```
Comment (schwartzmx) [inline packages/hyperion-autologging/src/SurfaceConsts.ts:7]: intentionally leaving off `id` suffix?

File: packages/hyperion-autologging/src/SurfaceProxy.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+   * separate container DOM node and shown in place later.
+   * Although DOM tree hierarchy is broken, the React Context hierarchy will
+   * continue to work. So, we use that fact and assign the right surface attribute
+   * to the container node. This way, when we walk up the DOM tree, we find the
+   * right surface value.
+   */
+  IReactDOMModule.createPortal.onArgsMapperAdd(args => {
+    const [_node, container] = args;
+    if (container instanceof HTMLElement) {
+      // args[0] = options.ReactModule.createElement(SurfaceProxy, { container }, node);
```
Comment (schwartzmx) [inline packages/hyperion-autologging/src/SurfaceProxy.ts:64]: could we bundle the changes here where we found the portal body container surface attribute mutating?

File: packages/hyperion-react/src/IReactComponent.ts [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+  function processReactFunctionComponent(functionComponent: TReactFunctionComponent): TReactFunctionComponent {
+    /**
+     * For functional components, we should always replace them with the
+     * intercepted version of them.
+     * however, the interceptFunction itself will only assign a FunctionIntercetpr
+     * to the function once.
+     * So, we don't need to use the interceptionInfo map here.
+     */
+
+    const fi = __DEV__
```
Comment (schwartzmx) [inline packages/hyperion-react/src/IReactComponent.ts:157]: will `__DEV__` work as expected here?

File: packages/hyperion-autologging/src/SurfaceConsts.ts (hunk 2 above)
Comment (reshadi) [reply packages/hyperion-autologging/src/SurfaceConsts.ts:7]: Yes, we should have done that long ago, I put the default without `id' but of course can be changed with a config.

File: packages/hyperion-autologging/src/SurfaceProxy.ts (hunk 3 above)
Comment (reshadi) [reply packages/hyperion-autologging/src/SurfaceProxy.ts:64]: I will have to pull in your fix from WWW and directly continue using the <SurfaceProxy>

File: packages/hyperion-react/src/IReactComponent.ts (hunk 4 above)
Comment (reshadi) [reply packages/hyperion-react/src/IReactComponent.ts:157]: Yes, I made sure of that, it works both separately if you use the hyperion-test-react-app and also when we pull in the code into WWW

**Review by schwartzmx — APPROVED** [review]
overall looks good,  accepting to unblock.  I think we know there are still a couple changes to port over.

**Review by vsylee — APPROVED** [review]

**Pull request merged**